service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,454 @@
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.math.BigDecimal;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Default implementation of {@link com.epam.esm.service.GiftCertificateService} interface.
+ *
+ * @since 1.0
+ */
+@Service
+@Transactional(propagation = Propagation.SUPPORTS, isolation = Isolation.READ_COMMITTED,
+        readOnly = true)
+public class GiftCertificateServiceImpl implements GiftCertificateService {
+
+    private final static String NAMES_PART_KEY = "namesPart";
+    private final static String DESCRIPTION_PART_KEY = "descriptionsPart";
+    private final static String TAG_NAMES_KEY = "tagNames";
+    private final static String SORT_FIELDS_KEY = "sortFields";
+    private final static String ORDER_KEY = "order";
+
+    private final ModelMapper modelMapper;
+    private final GiftCertificateDao giftCertificateDao;
+    private final TagDao tagDao;
+    private final CertificateTagsDao certificateTagsDao;
+
+    @Autowired
+    public GiftCertificateServiceImpl(GiftCertificateDao giftCertificateDao, ModelMapper modelMapper, TagDao tagDao,
+                                      CertificateTagsDao certificateTagsDao) {
+        this.giftCertificateDao = giftCertificateDao;
+        this.modelMapper = modelMapper;
+        this.tagDao = tagDao;
+        this.certificateTagsDao = certificateTagsDao;
+    }
+
+    /**
+     * This method do request to dao layer which depends on argument of the method.
+     * If the argument of the method contains @see {@link #NAMES_PART_KEY} than will be called {@link #getAllByPartOfNames(String[]) method}.
+     * If the argument of the method contains @see {@link #DESCRIPTION_PART_KEY} than will be called {@link #getAllByPartOfDescriptions(String[]) method}.
+     * If the argument of the method contains @see {@link #TAG_NAMES_KEY} than will be called {@link #getAllByTagNames(String[]) method}.
+     * If the argument of the method contains @see {@link #SORT_FIELDS_KEY} than will be called {@link #getAllSorted(String[], String)} method}.
+     *
+     * @param reqParams parameters of a request.
+     * @return List of GiftCertificateDao.
+     * @throws IllegalRequestParameterException if argument of the method contains {@link #SORT_FIELDS_KEY} key and
+     *                                          doesn't contain {@link #ORDER_KEY} key at the same time.
+     * @since 1.0
+     */
+    @Override
+    public List<GiftCertificateDto> findAllForQuery(Map<String, String[]> reqParams) {
+        for (Map.Entry<String, String[]> entry : reqParams.entrySet()) {
+            String key = entry.getKey();
+            switch (key) {
+                case NAMES_PART_KEY: {
+                    String[] namesPart = entry.getValue();
+                    return getAllByPartOfNames(namesPart);
```

Comment:
Is it possible to receive certificates searching by name and description and sort results at the same time?

Reply:
Remove all getAll*() methods. Replaced it by Criterias. Servica layer requests concrete Criteria with concrete Parameters(fields for searching) from CriteriaFactory and passes it to Dao layer. Sorting now is a part of Service layer and uses Comparators instead of ORDER BY sql queries. It can find entities by any fields and sorts the result on the Service layer.

commit cb8a000